### PR TITLE
[Feature] Family test list (EXPOSUREAPP-12321, EXPOSUREAPP-12327)

### PR DESCRIPTION
**Pull request by AlexanderAlferov**
### Whats new

Family test list screen.

Screenshot 2022-03-30 at 22 55 54

### Story

- [Story](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5895)
- [Family test list UI](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-12321)
- [Dismiss notification badge logic](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-12327)

### How to test

1. Take a cup of tea/coffee - this would be a long ride
2. Using INT env scan both PCR and RAT types for POSITIVE, NEGATIVE, INVALID
3. Go to family tests fragment and check, that every test appeared as it should be
4. Check all tests with notification badge
5. Navigate back to main fragment
6. Navigate back again to family tests fragment - check that all notification badges are gone
7. Switch to WRU env and check tests
8. Delete one test by swiping left and confirm in alert dialog
9. Delete one test by swiping right and confirm in alert dialog
10. Delete one test by pressing delete button (should be at least one in REDEEMED state)
11. Delete all left tests by using menu in top right corner (confirmation dialog should also appear)
12. Go to recycle bin and restore tests
13. Go to phone system settings and time travel (+10 years will be just fine)
14. Go back to family tests fragment - NEGATIVE PCR should became OUTDATED with delete button
15. Press that delete button
16. Delete all tests
17. Switch to Night Mode
18. Anger
19. Denial
20. Acceptance
21. Repeat steps 2-16 and check if everything is okay in Night Mode

### To Do

- [ ] Navigation to test details
- [x] Update Invalid and Outdated card icons
- [x] Update confirmation dialog text

### CODE REVIEW

**Review by CV113 — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt [hunk 1]
```diff
@@ -1,6 +1,98 @@
...
+    private val viewModel: FamilyTestListViewModel by cwaViewModels { viewModelFactory }
+    private val binding: FragmentFamilyTestListBinding by viewBinding()
+    private val familyTestListAdapter = FamilyTestListAdapter()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        setupMenu(binding.toolbar)
+        bindRecycler()
+        binding.refreshLayout.setOnRefreshListener { viewModel.onRefreshTests() }
+        binding.toolbar.setOnClickListener { viewModel.onBackPressed() }
```
[the remark this answers is not in the document]
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt:37]: also the toolbar title looks tight , can you please check?

File: Corona-Warn-App/src/main/res/values-de/family_tests_strings.xml [hunk 2]
```diff
@@ -37,4 +37,36 @@
     <!-- XBUT: Test registration selection: confirm button -->
     <string name="family_consent_cofirm_button">"Einverstanden"</string>
 
+    <!-- Family tests -->
+    <!-- XTXT: Family tests: Cards: PCR title -->
+    <string name="family_tests_cards_pcr_title">"PCR-Befund"</string>
```
Comment (KathrinLuecke) [inline Corona-Warn-App/src/main/res/values-de/family_tests_strings.xml:42]:
```suggestion
    <string name="family_tests_cards_pcr_title">"PCR-Test-Befund"</string>
```

**Review by KathrinLuecke — CHANGES_REQUESTED** [review]
see comment

File: Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    <TextView
+        android:id="@+id/title"
+        style="@style/headline6"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/card_padding"
+        android:layout_marginTop="24dp"
+        android:layout_marginEnd="@dimen/spacing_small"
+        android:focusable="false"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml:16]: is that required ? I see it in all texts in all items layouts

File: Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        app:layout_constrainedWidth="true"
+        app:layout_constraintHorizontal_bias="0"
+        app:layout_constraintEnd_toStartOf="@+id/icon"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:text="Mohamed Bernd Alexandru Juraj Chilja David Lukas" />
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml:22]: 😄 , it is a good reflection of team members at the moment and it is only visible for DEVs ;)

File: Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml (hunk 3 above)
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/res/layout/family_pcr_test_card_positiv.xml:16]: To be honest: I'm not sure - just used the same approach we used for submission cards, as it successfully passed accessibility tests 👀

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                }
+            } else {
+                familyTestRepository.moveTestToRecycleBin(test.identifier)
+            }
+        }
+    }
+
+    fun onRefreshTests() {
+        launch(appScope) {
+            familyTestRepository.refresh().also {
```
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            familyTestRepository.familyTests.first().map { familyTest ->
+                familyTestRepository.markBadgeAsViewed(familyTest.identifier)
+            }
+        }
+    }
+
+    fun onRemoveTestConfirmed(test: FamilyCoronaTest?) {
+        launch(appScope) {
+            if (test == null) {
+                familyTestRepository.familyTests.first().map { familyTest ->
```
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt [hunk 7]
```diff
@@ -1,6 +1,98 @@
...
+        viewModel.familyTests.observe2(this) { tests -> updateViews(tests) }
+        viewModel.events.observe2(this) { it?.let { onNavigationEvent(it) } }
+        viewModel.refreshComplete.observe2(this) { binding.refreshLayout.isRefreshing = false }
```
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt:42]: Done! ✅

File: Corona-Warn-App/src/main/res/values-de/family_tests_strings.xml (hunk 2 above)
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/res/values-de/family_tests_strings.xml]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt [hunk 8]
```diff
@@ -1,6 +1,98 @@
...
+            addItemDecoration(TopBottomPaddingDecorator(topPadding = R.dimen.spacing_normal))
+            setupSwipe(context = requireContext())
+        }
+    }
+
+    private fun showRemovalConfirmation(familyCoronaTest: FamilyCoronaTest?, position: Int?) =
+        MaterialAlertDialogBuilder(requireContext()).apply {
+            setTitle(
+                if (familyCoronaTest == null) R.string.family_tests_list_deletion_alert_header
+                else R.string.family_tests_list_deletion_alert_header
```
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListFragment.kt]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            } else {
+                familyTestRepository.moveTestToRecycleBin(test.identifier)
+            }
+        }
+    }
+
+    fun onRefreshTests() {
+        launch(appScope) {
+            familyTestRepository.refresh().also {
+                refreshComplete.postValue(null)
```
[the remark this answers is not in the document]
Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt:80]: Done! ✅

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            )
+            State.REDEEMED -> FamilyPcrTestRedeemedCard.Item(
+                familyCoronaTest = this,
+                onSwipeItem = { familyCoronaTest, position ->
+                    events.postValue(FamilyTestListEvent.ConfirmSwipeTest(familyCoronaTest, position))
+                },
+                onDeleteTest = { events.postValue(FamilyTestListEvent.ConfirmRemoveTest(this)) }
+            )
+            // Should not be possible
+            State.OUTDATED,
```
[the remark this answers is not in the document]
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/ui/testlist/FamilyTestListViewModel.kt:146]: ok

**Review by mtwalli — DISMISSED** [review]

**Comment by mtwalli** [comment]
@AlexanderAlferov , Very neat UI! 
Thanks for the hard work , you have done in CWA

**Review by jurajkusnier — DISMISSED** [review]

**Review by mtwalli — APPROVED** [review]

**Review by BMItr — APPROVED** [review]

**Review by KathrinLuecke — COMMENTED** [review]
UA approved

**Pull request merged**